st2 API and CLI command added for actions/workflows clone operation

mahesh-orch opened the pull request:
This PR adds st2 Clone API for actions and workflows.
1. API clones action from source pack to destination pack and renames them appropriately. 
2. Also, this operation makes changes in destination action metadata files like action name, pack name, entry point name, etc. Action clone API takes source_pack_name, source_action_name, destination_pack_name, and destination_action_name in the endpoint. 
3. The clone operation registers the new action/workflow to the database. The clone operation throws exceptions appropriately (i.e requested source does not exist, destination pack does not exist, destination action already exists unless overwrite is specified, a user does not have permission, etc.) 
4. The clone operation takes an optional parameter in the query (?overwrite=True) to force the clone (or overwrite if the destination already exists).
5. CLI command added for cloning operation. It takes four positional arguments source_pack_name, source_action_name, destination_pack_name and destination_action_name. Also, it takes an optional argument (`-f` or `--force`) to overwrite action file content if the destination already exists.

m4dcoder commented on st2common/st2common/validators/api/misc.py:
```diff
@@ -37,3 +37,9 @@ def validate_not_part_of_system_pack(resource_db):
+
+
+def validate_not_part_of_system_pack_by_name(pack_name):
+    if pack_name == SYSTEM_PACK_NAME:
```
Should check that it's not in SYSTEM_PACK_NAMES (list).

m4dcoder commented on st2common/tests/unit/services/test_packs.py:
```diff
@@ -241,3 +257,148 @@ def test_exception_to_remove_resource_metadata_file(self, remove):
+                TEST_DEST_PACK_PATH,
+                TEST_DEST_PACK,
+                "action_6",
+            )
```
Please add unit tests for RBAC test cases.

m4dcoder commented on st2common/st2common/services/packs.py:
```diff
@@ -290,3 +291,95 @@ def delete_action_files_from_pack(pack_name, entry_point, metadata_file):
+        source_entry_point_file_path = os.path.join(
+            source_pack_base_path, "actions", source_entry_point
+        )
+        if str(source_entry_point).endswith("py"):
```
Should be using the runner type in the metadata file to determine if this is python action.

m4dcoder commented on st2common/st2common/services/packs.py:
```diff
@@ -290,3 +291,95 @@ def delete_action_files_from_pack(pack_name, entry_point, metadata_file):
+        )
+        if str(source_entry_point).endswith("py"):
+            dest_entry_point_file_name = "%s.py" % (dest_action)
+        elif str(source_entry_point).startswith("workflows"):
```
Should be using the runner type in the metadata file to determine if this is a workflow.

m4dcoder commented on st2common/st2common/services/packs.py:
```diff
@@ -290,3 +291,95 @@ def delete_action_files_from_pack(pack_name, entry_point, metadata_file):
+        with open(source_file, "r") as sf:
+            with open(destination_file, "w") as df:
+                for line in sf:
+                    df.write(line)
```
Can we use shutil here?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/StackStorm/st2?pullRequest=5345) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/StackStorm/st2?pullRequest=5345) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/StackStorm/st2?pullRequest=5345) it.</sub>

mahesh-orch commented:
@m4dcoder 
* Permission type changed as ACTION_VIEW on source action, ACTION_CREATE on destination action. Added unit test for each scenario (https://github.com/StackStorm/st2-rbac-backend/pull/56).
* Here for registering single cloned action there are two options:
	1. To copy the action the source_action_db instance to destination_action_db and modify it according to the user inputs and then use `Action.add_or_update(destination_action_db)` to register it to database. If destination already exists then de-register it before and then register a cloned_destination_action_db.
(I have used the first way) Also, it is convenient here to check ACTION_CREATE permission on destination pack if destination action doesn't exist.
   	2. To write a function in /st2common/bootstrap/actionsregistrar.py to register single action in a pack.
* I didn't consolidate logic to single self.manager.clone because for args.force or user input "y" or "yes" overwrite flag will become True otherwise False and it is being used to prepare payload as body or clone post request.
* As suggested, implemented ActionCloneCommand and returning NotImplementedError in ResourceCloneCommand for other resources.
* Changed clone API endpoint to `/api/v1/actions/{ref_or_id}` and accepting other required and optional args in json body.
* Added a function named `remove_unnecessary_files_from_pack` in /st2common/services/packs.py. While cloning an action, if there is a case of overwrite destination and the source runner type is different than the destination runner type then after cloning operation, unnecessary entry point files need to be removed from destination pack. 
I have added a separate function for this purpose.
* Also, have made other minor changes as suggested.
Please review the changes. Thanks.

mahesh-orch commented:
@m4dcoder 
* Permission type changed as ACTION_VIEW on source action, ACTION_CREATE on destination action. Added unit test for each scenario (https://github.com/ashwini-orchestral/st2/blob/clone_action_api/st2api/tests/unit/controllers/v1/test_actions_rbac.py)
* Here for registering single cloned action there are two options:
	1. To copy the action the source_action_db instance to destination_action_db and modify it according to the user inputs and then use `Action.add_or_update(destination_action_db)` to register it to database. If destination already exists then de-register it before and then register a cloned_destination_action_db.
(I have used the first way) Also, it is convenient here to check ACTION_CREATE permission on destination pack if destination action doesn't exist.
   	2. To write a function in /st2common/bootstrap/actionsregistrar.py to register single action in a pack.
* I didn't consolidate logic to single self.manager.clone because for args.force or user input "y" or "yes" overwrite flag will become True otherwise False and it is being used to prepare payload as body or clone post request.
* As suggested, implemented ActionCloneCommand and returning NotImplementedError in ResourceCloneCommand for other resources.
* Changed clone API endpoint to `/api/v1/actions/{ref_or_id}` and accepting other required and optional args in json body.
* Added a function named `remove_unnecessary_files_from_pack` in /st2common/services/packs.py. While cloning an action, if there is a case of overwrite destination and the source runner type is different than the destination runner type then after cloning operation, unnecessary entry point files need to be removed from destination pack. 
I have added a separate function for this purpose.
* Also, have made other minor changes as suggested.
Please review the changes. Thanks.

m4dcoder commented on st2client/st2client/models/core.py:
```diff
@@ -376,6 +376,31 @@ def delete(self, instance, **kwargs):
+        url = "/%s/%s/clone/?overwrite=%s" % (
+            self.resource.get_url_path_name(),
+            source_ref,
+            overwrite,
```
Overwrite is already in the JSON payload. Is this duplicate? Let's just use the overwrite in the payload below.

m4dcoder commented on st2api/st2api/controllers/v1/actions.py:
```diff
@@ -271,6 +275,134 @@ def delete(self, ref_or_id, requester_user):
+                dest_pack_base_path=dest_pack_base_path,
+                dest_pack=dest_data.dest_pack,
+                dest_action=dest_data.dest_action,
+            )
```
Catch permission error and abort with unauthorized.

mahesh-orch commented:
@m4dcoder 
* I have refactored the clone action function as per your suggestions and refactored corresponding unit tests.
* I kept clone_action_files and clone_action_db functions separate, because the destination action_db will not be available before cloning files to check ACTION_CREATE permission on destination pack.
* Used delete delete API to remove action from db and disk in case of overwrite.
* Removed entry point file delete function in case of source and destination action runners are not same.
* Also, done other minor changes as said.
Please review. Thanks.

mahesh-orch commented:
@m4dcoder 
* I have refactored the clone action function as per your suggestions and refactored corresponding unit tests.
* I kept clone_action_files and clone_action_db functions separate, because the destination action_db will not be available before cloning files to check ACTION_CREATE permission on destination pack.
* Used delete delete API to remove action from db and disk in case of overwrite.
* Removed entry point file delete function in case of source and destination action runners are not same.
* Also, done other minor changes as said.
Please review. Thanks.
I couldn't resolve why CI integration check is failing.

m4dcoder commented on st2api/st2api/controllers/v1/actions.py:
```diff
@@ -281,6 +285,126 @@ def delete(self, ref_or_id, requester_user):
+            LOG.error(
+                "Exception encountered during cloning action. Exception was %s",
+                e,
+            )
```
Clean up and remove cloned files if the clone operation is not successful.

m4dcoder commented on st2api/tests/unit/controllers/v1/test_actions.py:
```diff
@@ -809,6 +843,165 @@ def test_delete_action_belonging_to_system_pack(self):
+    @mock.patch.object(
+        action_validator, "validate_action", mock.MagicMock(return_value=True)
+    )
+    def test_clone_overwrite_exception_cloning_db_action_reregistered(
```
Rename this to `test_clone_overwrite_exception_source_recovered`

m4dcoder commented on st2api/tests/unit/controllers/v1/test_actions.py:
```diff
@@ -809,6 +843,165 @@ def test_delete_action_belonging_to_system_pack(self):
+        self.assertEqual(clone_resp.json["faultstring"], msg)
+
+        self.__do_delete(self.__get_action_id(source_post_resp))
+
```
Check that the source action is reregistered.

m4dcoder commented on st2api/tests/unit/controllers/v1/test_actions.py:
```diff
@@ -809,6 +843,165 @@ def test_delete_action_belonging_to_system_pack(self):
+        source_post_resp = self.__do_post(ACTION_16)
+        self.assertEqual(source_post_resp.status_int, 201)
+        dest_post_resp = self.__do_post(ACTION_17)
+        self.assertEqual(dest_post_resp.status_int, 201)
```
Why are we creating ACTION_17 here? Should clone do that?

m4dcoder commented on st2api/tests/unit/controllers/v1/test_actions.py:
```diff
@@ -809,6 +843,165 @@ def test_delete_action_belonging_to_system_pack(self):
+        }
+        source_ref_or_id = "%s.%s" % (ACTION_16["pack"], ACTION_16["name"])
+        clone_resp = self.__do_clone(dest_data_body, source_ref_or_id)
+        self.assertEqual(clone_resp.status_int, 201)
```
Check ACTION_17 exists

m4dcoder commented on st2api/tests/unit/controllers/v1/test_actions.py:
```diff
@@ -809,6 +843,165 @@ def test_delete_action_belonging_to_system_pack(self):
+        source_post_resp = self.__do_post(ACTION_16)
+        self.assertEqual(source_post_resp.status_int, 201)
+        dest_post_resp = self.__do_post(ACTION_17)
+        self.assertEqual(dest_post_resp.status_int, 201)
```
Create an ACTION_17 with something different so we can test below that we overwrote it with something else.

m4dcoder commented on st2api/tests/unit/controllers/v1/test_actions.py:
```diff
@@ -809,6 +843,165 @@ def test_delete_action_belonging_to_system_pack(self):
+        }
+        source_ref_or_id = "%s.%s" % (ACTION_16["pack"], ACTION_16["name"])
+        clone_resp = self.__do_clone(dest_data_body, source_ref_or_id)
+        self.assertEqual(clone_resp.status_int, 201)
```
Check ACTION_17 exists and that it's different from the original

m4dcoder commented on st2api/tests/unit/controllers/v1/test_actions.py:
```diff
@@ -809,6 +843,165 @@ def test_delete_action_belonging_to_system_pack(self):
+        self.assertEqual(clone_resp.status_int, 500)
+        self.assertEqual(clone_resp.json["faultstring"], msg)
+        self.__do_delete(self.__get_action_id(source_post_resp))
+        self.__do_delete(self.__get_action_id(dest_post_resp))
```
Add test on clone exception with file Permission Error

m4dcoder commented on st2api/tests/unit/controllers/v1/test_actions.py:
```diff
@@ -809,6 +843,165 @@ def test_delete_action_belonging_to_system_pack(self):
+            expect_errors=True,
+        )
+        self.assertEqual(clone_resp.status_int, 500)
+        self.assertEqual(clone_resp.json["faultstring"], msg)
```
Test that any files that were cloned are cleaned up.

m4dcoder commented on st2client/st2client/commands/action.py:
```diff
@@ -317,6 +318,92 @@ def run_and_print(self, args, **kwargs):
+                dest_action,
+                overwrite=False,
+                **kwargs,
+            )
```
Clean up this code block and deduplicate `self.manager.clone`

m4dcoder commented on st2common/st2common/services/packs.py:
```diff
@@ -290,3 +293,116 @@ def delete_action_files_from_pack(pack_name, entry_point, metadata_file):
+        elif source_runner_type == "orquesta":
+            dest_entry_point_file_name = "workflows/%s.yaml" % (dest_action)
+        else:
+            dest_entry_point_file_name = dest_action
```
- This looks duplicate from the clone_action_db. Since we ran clone_action_db before, can we take the entry point from the dest_action here?
- Check if the workflows directory exist and create if it does not

m4dcoder commented on st2common/tests/unit/services/test_packs.py:
```diff
@@ -241,3 +257,156 @@ def test_exception_to_remove_resource_metadata_file(self, remove):
+        expected_msg = (
+            "Please make sure 'workflows' directory present in path: '%s'"
+            % (cloned_action_entry_point_file_path)
+        )
```
clone_action should create the workflows directory if it doesn't exist

m4dcoder commented on st2common/tests/unit/services/test_packs.py:
```diff
@@ -241,3 +257,156 @@ def test_exception_to_remove_resource_metadata_file(self, remove):
+                TEST_DEST_PACK,
+                "action_5",
+            )
+
```
- Add a unit test to check if destination pack does not exists.
- Add a unit test to check if destination files already exist.

m4dcoder commented on st2api/tests/unit/controllers/v1/test_actions_rbac.py:
```diff
@@ -0,0 +1,392 @@
+# Copyright 2020 The StackStorm Authors.
```
We will have to move these tests to https://github.com/StackStorm/st2-rbac-backend/blob/master/tests/unit/controllers/api/v1/test_actions_rbac.py. You will need to open a separate PR at the st2-rbac-backend repo. Let's do this when the code is ready so we know the changes are working.

ashwini-orchestral commented:
@m4dcoder 
 * I have refactored the action clone method according to the pseudo code you provided.
 * Used action create API and action delete API.
 * I have refactored existing unit tests as well as added new unit tests as you suggested.
 * Renamed unit test name (https://github.com/StackStorm/st2/pull/5345/files#r709666222) as `test_clone_overwrite_exception_destination_recovered` as we are deleting destination action along with files if it exists and overwrite flag is true.
 * Clean up code in st2client/commands/action.py and did deduplicate `self.manager.clone`.
 * Didn't clean up as per comment https://github.com/StackStorm/st2/pull/5345/files#r709669852,  to keep `clone_actions` function testable.
 * Checking for workflows directory exist, if it isn't then creating it. Added separate unit test to this change.
 * Also, done other minor fixes
 * Wanted to inform that the existing workflows directory has permission mode 775 and if it doesn't exist and we create it then it has permission mode 731 even if I mention it 775 while creating directory.
Please review the changes. Thanks.

mahesh-orch commented:
@m4dcoder 
 * I have refactored the action clone method according to the pseudo code you provided.
 * Used action create API and action delete API.
 * I have refactored existing unit tests as well as added new unit tests as you suggested.
 * Renamed unit test name (https://github.com/StackStorm/st2/pull/5345/files#r709666222) as `test_clone_overwrite_exception_destination_recovered` as we are deleting destination action along with files if it exists and overwrite flag is true.
 * Clean up code in st2client/commands/action.py and did deduplicate `self.manager.clone`.
 * Didn't clean up as per comment https://github.com/StackStorm/st2/pull/5345/files#r709669852,  to keep `clone_actions` function testable.
 * Checking for workflows directory exist, if it isn't then creating it. Added separate unit test to this change.
 * Also, done other minor fixes
 * Wanted to inform that the existing workflows directory has permission mode 775 and if it doesn't exist and we create it then it has permission mode 731 even if I mention it 775 while creating directory.
Please review the changes. Thanks.

mahesh-orch commented:
@m4dcoder 
 * I have refactored the action clone method according to the pseudo code you provided.
 * Used action create API and action delete API.
 * I have refactored existing unit tests as well as added new unit tests as you suggested.
 * Renamed unit test name as `test_clone_overwrite_exception_destination_recovered` under st2api/tests/unit/controller/v1/test_actions.py, as we are deleting destination action along with files if it exists and overwrite flag is true.
 * Clean up code in st2client/commands/action.py and did deduplicate `self.manager.clone`.
 * Didn't clean up under st2common/services/packs.py to use `dest_action_db` to retrieve dest_entry_point and dest_metadata_file as to keep `clone_actions` function testable.
 * Checking for workflows directory exist, if it isn't then creating it. Added separate unit test to this change.
 * Also, done other minor fixes
 * Wanted to inform that the existing workflows directory has permission mode 775 and if it doesn't exist and we create it then it has permission mode 731 even if I mention it 775 while creating directory.
Please review the changes. Thanks.

cognifloyd commented on st2client/st2client/commands/action.py:
```diff
@@ -307,6 +308,92 @@ def run_and_print(self, args, **kwargs):
+        except ResourceNotFoundError:
+            dest_instance = None
+
+        if dest_instance:
+            user_input = ""
+            if not args.force:
+                user_input = input(
+                    "The destination action already exists. Do you want to overwrite? (y/n): "
+                )
+            if args.force or user_input.lower() == "y" or user_input.lower() == "yes":
+                return self.manager.clone(
+                    source_ref,
+                    dest_pack,
+                    dest_action,
+                    overwrite=True,
+                    **kwargs,
+                )
+            else:
+                print("Action is not cloned.")
+                return
+        else:
+            return self.manager.clone(
+                source_ref,
+                dest_pack,
+                dest_action,
+                overwrite=False,
+                **kwargs,
+            )
+
```
```suggestion
        overwrite = False

        if dest_instance:
            user_input = ""
            if not args.force:
                user_input = input(
                    "The destination action already exists. Do you want to overwrite? (y/n): "
                )
            if args.force or user_input.lower() == "y" or user_input.lower() == "yes":
                overwrite = True
            else:
                print("Action is not cloned.")
                return

        return self.manager.clone(
            source_ref,
            dest_pack,
            dest_action,
            overwrite=overwrite,
            **kwargs,
        )

```

cognifloyd reviewed (commented):
I have one suggestion. What else is left before we can merge this?

cognifloyd commented:
Hmm. looks like we at least need the https://github.com/StackStorm/st2-rbac-backend PR for the tests for this.
Also, @m4dcoder have your concerns been addressed here?

I'm thinking this needs to be bumped to 3.7.0.

m4dcoder commented on st2common/st2common/services/packs.py:
```diff
@@ -290,3 +293,112 @@ def delete_action_files_from_pack(pack_name, entry_point, metadata_file):
+    dest_action_db["name"] = dest_action
+    dest_ref = ".".join([dest_pack, dest_action])
+    dest_action_db["ref"] = dest_ref
+    dest_action_db["uid"] = "action:%s:%s" % (dest_pack, dest_action)
```
Instead of using colon, use the UID separator `UIDFieldMixin.UID_SEPARATOR` and import `from st2common.models.db.stormbase import UIDFieldMixin`. If there's any change to the separator symbol, this won't be broken.

```
```

m4dcoder commented on st2common/st2common/services/packs.py:
```diff
@@ -290,3 +293,112 @@ def delete_action_files_from_pack(pack_name, entry_point, metadata_file):
+        yaml.dump(doc, df, default_flow_style=False, sort_keys=False)
+
+
+def clone_action_db(source_action_db, dest_pack, dest_action):
```
I don't see any unit tests for this function. Please point me to them.

mahesh-orch commented:
@m4dcoder I have addressed your review comments. Please review. Thanks.

armab commented:
@mahesh-orch @m4dcoder Thanks for the massive work here!
The PR is missing the Changelog. Please add it in another PR so we'll track record of the changes around the st2.

mahesh-orch commented:
@armab Sure, I will create new PR for adding CHANGELOG. Thanks.
